wcomponents-core/src/main/java/com/github/bordertech/wcomponents/WFileWidget.java

```diff
@@ -110,13 +133,109 @@ protected boolean doHandleRequest(final Request request) {
+			boolean validFileType;
+			if (isSetFileTypes()) {
+				validFileType = FileValidationUtil.validateFileType(value, getFileTypes());
+				addFileTypeValidation(validFileType, InternalMessages.DEFAULT_VALIDATION_ERROR_FILE_WRONG_TYPE, 
+						getFileTypes().toArray(new Object[getFileTypes().size()]));
+			} else {
+				validFileType = true;
+			}
+			
+			// if fileSize is supplied then validate it
+			boolean validFileSize;
+			if (isSetFileSize()) {
+				validFileSize = FileValidationUtil.validateFileSize(value, getMaxFileSize());
+				addFileSizeValidation(validFileSize, InternalMessages.DEFAULT_VALIDATION_ERROR_FILE_WRONG_SIZE, 
+						Util.readableFileSize(value.getSize()), Util.readableFileSize(getMaxFileSize()));
+			} else {
+				validFileSize = true;
+			}
+			
+			// if file is valid, the update data
+			if (validFileSize && validFileType) {
+				setData(value);
+			} else {
+				// otherwise no change
+				changed = false;
+			}
+		} 
 
 		return changed;
 	}
 
 	/**
+	 * Reset validation state.
+	 */
+	public void resetValidationState() {
+		getOrCreateComponentModel().validationMessages.clear();
+		getOrCreateComponentModel().validFileType = null;
+		getOrCreateComponentModel().validFileSize = null;
+	}
+
+	/**
+	 * Adds invalid file type validation message.
+	 * @param fieldValue True/False
+	 * @param message The message to add.
+	 * @param args The message arguments.
+	 */
+	private void addFileTypeValidation(final boolean fieldValue, final String message, final Object... args) {
+		getOrCreateComponentModel().validFileType = fieldValue;
+		if (!fieldValue) {
+			String invalidMessage = String.format(I18nUtilities.format(null, message), StringUtils.join(args, ","));
+			getOrCreateComponentModel().validationMessages.add(invalidMessage);
+		}
+	}
+
+	/**
+	 * Adds invalid file size validation message.
+	 * @param fieldValue True/False
+	 * @param message The message to add.
+	 * @param args The message arguments.
+	 */
+	private void addFileSizeValidation(final boolean fieldValue, final String message, final Object... args) {
+		getOrCreateComponentModel().validFileSize = fieldValue;
+		if (!fieldValue) {
+			String invalidMessage = String.format(I18nUtilities.format(null, message), args);
+			getOrCreateComponentModel().validationMessages.add(invalidMessage);
+		}
+	}
+
+	/**
+	 * Indicates whether the uploaded file is valid, if {@link getFileTypes()} is set.
+	 *
+	 * @return true if file type valid, false file type invalid, otherwise null.
+	 * @see {@link getFileValidationMessages()}
+	 */
+	public Boolean isFileTypeValid() {
+		return getComponentModel().validFileType;
+	}
+
+	/**
+	 * Indicates whether the uploaded file is valid, if {@link getMaxFileSize()} is set.
+	 *
+	 * @return true if file size valid, false file size invalid, otherwise null.
+	 * @see {@link getFileValidationMessages()}
+	 */
+	public Boolean isFileSizeValid() {
+		return getComponentModel().validFileSize;
+	}
+
```

Comment:
ValidateComponent will also need to be overridden to include the new validation messages.